### ISSUE TITLE
Video Optimization: Improve default ffmpeg configuration

### ISSUE THREAD

**Issue by swissspidy** [GoogleForCreators/web-stories-wp#5993]
## Feature Description

Let's try to optimize the default ffmpeg configuration to get the best speed/quality/size combo.

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Implementation Brief

* [ ] See how the ffmpeg configuration could be optimized, and do that

**Comment by swissspidy** [comment]
@pbakaus IIRC you are a bit of an ffmpeg expert by now. Any pointers how to get the best out of ffmpeg for within the editor?

Currently we use:

```
ffmpeg -i inputfile.mp4 -vcodec libx264 -vf "scale='min(1080,iw)':'min(1920,ih)'" -preset fast outputfile.mp4
```

I've also tried `-preset veryfast` before but that seems to crash in the browser.

**Comment by pbakaus** [comment]
mhh the basics here seem correct to me. Weird that veryfast/ultrafast would crash.. See https://trac.ffmpeg.org/wiki/Encode/H.264 for more, I would also try:

- I don't see any quality setting - try `-crf`
- try `zerolatency` and `-movflags +faststart`
- I'm guessing OpenCL / HW accel won't work in webasm ;)
- try a dumber color profile: `-vf format=yuv420p`
- try a simpler codec profile: `-profile:v baseline` or `-profile:v main`. This will will disable a bunch of advanced optimization strategies that make encoding/decoding slower.

Also, try `veryfast` again with the color/codec profile changes, could make a difference.

**Comment by spacedmonkey** [comment]
@swissspidy Do you want to put the file as a webm instead of mp4?

**Comment by spacedmonkey** [comment]
This [table](https://en.wikipedia.org/wiki/HTML5_video#Browser_support) is interesting from support.

**Comment by swissspidy** [comment]
> @swissspidy Do you want to put the file as a webm instead of mp4?

No, the format recommendation is to use MP4 for best compatibility. So we stick to MP4.

**Comment by tokumine** [comment]
drive-by: did folks try HLS or are the costs of implementing both HLS and MP4 (for desktop fallback) prohibitive (CPU/Storage)?

**Comment by swissspidy** [comment]
> drive-by: did folks try HLS or are the costs of implementing both HLS and MP4 (for desktop fallback) prohibitive (CPU/Storage)?

Storage costs are one factor, limited browser support and complexity of storing multiple formats within WordPress are others. It's something we might explore in the future if the benefits outweigh the drawbacks.

**Comment by campiotto** [comment]
@swissspidy can't wait to start using the hotlink feature, could you let me know what the current ffmpeg configuration is the plugin is using?

I want to keep your coding pattern.

**Comment by swissspidy** [comment]
@campiotto The current configuration can be found at https://github.com/google/web-stories-wp/blob/eaf67db8d83482b2ca027eda019aa79c0df72bb3/packages/story-editor/src/app/media/utils/useFFmpeg.js, with some explanation available at https://pascalbirchler.com/client-side-video-optimization/. This of course is irrelevant for hotlinking as we can't run hotlinked videos through FFmpeg.

**Comment by campiotto** [comment]
@swissspidy I think there is some problem with using hotlink. The test I took was as follows:

I used 4 media videos that had been generated by the plugin, moved them to my CDN, then added them back to webstories via hotlink.

I didn't change anything in the encoding of these videos, the plugin who did everything I just moved to the CDN.

In the plugin settings I always had google cache enabled, however the image quality was poor using hotlink and with this the view rate dropped a lot.
